csi/moac/volume.ts

```diff
@@ -0,0 +1,877 @@
+      const replica = this.replicas[nodeName];
+      if (replica.isOffline()) {
+        continue;
+      }
+      let share;
+      const isLocal = replica.pool!.node.name === localNode;
+      if (isLocal && replica.share !== 'REPLICA_NONE') {
+        // make sure that replica that is local to the nexus is accessed locally
+        share = 'REPLICA_NONE';
+      } else if (!isLocal && replica.share === 'REPLICA_NONE') {
+        // make sure that replica that is remote to nexus can be accessed
+        share = 'REPLICA_NVMF';
+      }
+      if (share) {
+        try {
+          await replica.setShare(share);
+          delete this.nodeBlackList[nodeName];
+          // fsa will get called again because the replica was modified
+          return;
+        } catch (err) {
+          this.nodeBlackList[nodeName] = true;
+          log.error(
+            `Failed to set share protocol to ${share} for replica "${replica}": ${err}`
+          );
+        }
+      }
+    }
+
+    // If we don't have a nexus and the volume is published, then try to create one
+    if (!this.nexus) {
+      assert(this.publishedOn);
+      let nexusNode = this.registry.getNode(this.publishedOn);
+      if (nexusNode && nexusNode.isSynced()) {
+        let replicas = [];
+        for (let nodeName in this.replicas) {
+          if (!this.replicas[nodeName].isOffline() && !this.nodeBlackList[nodeName]) {
+            replicas.push(this.replicas[nodeName]);
+          }
+        }
+        if (replicas.length === 0) {
+          log.warn(`Cannot create nexus for ${this} because all replicas are bad`);
+          return;
+        }
+        try {
+          await this._createNexus(nexusNode, replicas);
+        } catch (err) {
+          log.error(`Failed to create nexus for ${this} on "${this.publishedOn}"`);
+          this._setState(VolumeState.Faulted);
+        }
+      } else {
+        log.warn(`Cannot create nexus for ${this} because "${this.publishedOn}" is down`)
+        this._setState(VolumeState.Faulted);
+      }
+      // fsa will get called again when event about created nexus arrives
+      return;
+    }
+
+    // pair nexus children with replica objects to get the full picture
+    const childReplicaPairs: {ch: Child, r: Replica | undefined}[] = this.nexus.children.map((ch) => {
+      const r = Object.values(this.replicas).find((r) => r.uri === ch.uri);
+      return {ch, r};
+    });
+    // add newly found replicas to the nexus (one by one)
+    const newReplicas = Object.values(this.replicas).filter((r) => {
+      return (!r.isOffline() &&
+        !childReplicaPairs.find((pair) => pair.r === r) &&
+        !this.nodeBlackList[r.pool!.node!.name]);
+    });
+    for (let i = 0; i < newReplicas.length; i++) {
+      try {
+        await this.nexus.addReplica(newReplicas[i]);
+        return;
+      } catch (err) {
+        // XXX what should we do with the replica? Destroy it?
+        this.nodeBlackList[newReplicas[i].pool!.node!.name] = true;
```

Comment:
I guess this can happen if the node where the replica lives has just gone offline, but it can also happen if there is something wrong with the replica itself. Probably not much else we can do at the moment...

Reply:
yea. I have introduced nodeBlackList to avoid the most obvious problems of endless loops when trying to do something with replica, it does not work, but next time we are in FSA we do the same thing again and cannot move forward. So it's important feature, however it did not receive any testing (both manual and unit-tests) so later I will probably have to come back to it and make the feature complete if it does not work according to expectations.